File: homing-workspace/src/main/resources/homing/js/hue/captains/singapura/js/homing/workspace/WidgetPickerModule.js
// =============================================================================
// WidgetPickerModule — RFC 0025 Ext1b Mechanism 2.
//
// Ephemeral picker UI. One instance per "add widget" action; once the user
// picks (or cancels), the picker goes out of scope. The picker holds NO
// state, owns NO branches, knows NOTHING about lifecycle, and renders into
// a caller-supplied host element — caller decides the container (a tab
// content area, a modal body, a sidebar, …).
//
//   var picker = new WidgetPicker({
//       entries     : WidgetEntryJson[],            // already filtered by caller
//       disabledIds : { "MyWidget": true },         // optional — greys those tiles
//       onPick      : function(entry, params) {},   // fired exactly once on commit
//       onCancel    : function() {}                 // optional — fired on form Cancel
//   });
//   picker.mountInto(hostEl);
//
// The orchestration of branch creation, tab placement, dynamic import, and
// tab-close cleanup lives in the workspace's chrome widget. The picker is
// intentionally dumb.
// =============================================================================

// Picker styles live in WidgetPickerStyles (CssGroup) — imported by the
// Java module declaration so the css.* identifiers are in scope here.

class WidgetPicker {

    constructor(opts) {
        if (!opts || !Array.isArray(opts.entries)) {
            throw new Error("WidgetPicker: entries[] required");
        }
        if (typeof opts.onPick !== "function") {
            throw new Error("WidgetPicker: onPick(entry, params) required");
        }
        this._entries     = opts.entries;
        this._disabledIds = opts.disabledIds || {};
        this._onPick      = opts.onPick;
        this._onCancel    = opts.onCancel || null;
        this._host        = null;
        this._delivered   = false;     // ensures onPick fires at most once
    }

    /**
     * Render the tile grid into hostEl. Caller chooses the container —
     * empty tab body, modal body, sidebar, anywhere. Picker takes over
     * the host's content (replaceChildren) but doesn't own the host
     * lifecycle. On pick or cancel the picker's content is replaced by
     * the caller (typically with the widget root or by removing the
     * tab entirely).
     */
    mountInto(hostEl) {
        this._host = hostEl;
        hostEl.replaceChildren(this._buildGrid());
    }

    // ─── Grid + tiles ──────────────────────────────────────────────────────

    _buildGrid() {
        var root = document.createElement("div");
        css.setClass(root, hwp_grid);

        // Group entries by entry.group, preserving registration order.
        var groups = [];
        var idx = {};
        for (var i = 0; i < this._entries.length; i++) {
            var e = this._entries[i];
            if (!(e.group in idx)) { idx[e.group] = groups.length; groups.push({ name: e.group, items: [] }); }
            groups[idx[e.group]].items.push(e);
        }

        for (var g = 0; g < groups.length; g++) {
            var label = document.createElement("div");
            css.setClass(label, hwp_group_label);
            label.textContent = groups[g].name;
            root.appendChild(label);
            for (var k = 0; k < groups[g].items.length; k++) {
                root.appendChild(this._buildTile(groups[g].items[k]));
            }
        }
        return root;
    }

    _buildTile(entry) {
        var tile = document.createElement("div");
        css.setClass(tile, hwp_tile);
        var disabled = !!this._disabledIds[entry.simpleName];
        if (disabled) {
            css.addClass(tile, hwp_tile_disabled);
            tile.title = "Already open";
        }

        var icon = document.createElement("div");
        css.setClass(icon, hwp_tile_icon);
        icon.textContent = entry.icon && entry.icon.kind === "emoji" ? entry.icon.value : "📦";
        tile.appendChild(icon);

        var label = document.createElement("div");
        css.setClass(label, hwp_tile_label);
        label.textContent = entry.label;
        tile.appendChild(label);

        if (entry.description) {
            var desc = document.createElement("div");
            css.setClass(desc, hwp_tile_desc);
            desc.textContent = entry.description;
            tile.appendChild(desc);
        }

        var self = this;
        tile.addEventListener("click", function () {
            if (disabled) {
                // Disabled-tile click means "I want this one but it's open" —
                // signal the caller via params=null so it can focus the
                // existing instance.
                if (self._delivered) return;
                self._delivered = true;
                self._onPick(entry, null);
                return;
            }
            self._onTilePick(entry);
        });
        return tile;
    }

    // ─── Pick handling ─────────────────────────────────────────────────────

    _onTilePick(entry) {
        var hasParams = entry.paramsFields && entry.paramsFields.length > 0;
        if (!hasParams) {
            this._deliver(entry, {});
        } else {
            this._showForm(entry);
        }
    }

    _showForm(entry) {
        var self = this;
        var defs = entry.defaults || {};
        var form = document.createElement("div");
        css.setClass(form, hwp_form);
        var inputs = {};
        for (var i = 0; i < entry.paramsFields.length; i++) {
            var f = entry.paramsFields[i];
            var row = document.createElement("div");
            css.setClass(row, hwp_form_row);
            var lab = document.createElement("label");
            css.setClass(lab, hwp_form_label);
            lab.textContent = f.name + (f.type ? " (" + f.type + ")" : "");
            var inp = document.createElement("input");
            css.setClass(inp, hwp_form_input);
            inp.value = defs[f.name] != null ? defs[f.name] : "";
            row.appendChild(lab); row.appendChild(inp);
            form.appendChild(row);
            inputs[f.name] = inp;
        }
        var actions = document.createElement("div");
        css.setClass(actions, hwp_form_actions);
        var cancelBtn = document.createElement("button");
        css.setClass(cancelBtn, hwp_form_btn);
        cancelBtn.textContent = "Cancel";
        var okBtn = document.createElement("button");
        css.setClass(okBtn, hwp_form_btn);
        css.addClass(okBtn, hwp_form_btn_primary);
        okBtn.textContent = "Open";
        actions.appendChild(cancelBtn); actions.appendChild(okBtn);
        form.appendChild(actions);

        if (this._host) this._host.replaceChildren(form);

        cancelBtn.addEventListener("click", function () {
            if (self._delivered) return;
            self._delivered = true;
            if (self._onCancel) self._onCancel();
        });
        okBtn.addEventListener("click", function () {
            var params = {};
            for (var k in inputs) {
                if (inputs.hasOwnProperty(k)) params[k] = inputs[k].value;
            }
            self._deliver(entry, params);
        });
    }

    _deliver(entry, params) {
        if (this._delivered) return;
        this._delivered = true;
        this._onPick(entry, params);
    }
}

File: homing-workspace/src/main/java/hue/captains/singapura/js/homing/workspace/WidgetPickerModule.java
package hue.captains.singapura.js.homing.workspace;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;
import hue.captains.singapura.js.homing.core.ModuleImports;

import java.util.List;

/**
 * RFC 0025 Ext1b Mechanism 2 — the widget picker UI.
 *
 * <p>Vends the {@code WidgetPicker} JS class: an <b>ephemeral, stateless</b>
 * tile grid + optional params form that renders into a caller-supplied
 * host element. One instance per "add widget" intent; goes out of scope
 * after pick or cancel. Holds no state, owns no branches, knows nothing
 * about lifecycle — pure UI.</p>
 *
 * <h2>Wiring contract</h2>
 *
 * <p>The chrome creates a fresh picker per add-intent and mounts it into
 * any host element it controls (typically the content area of an empty
 * tab spawned by {@code MultiTabPane.onAddTab}; could also be a modal
 * body or a sidebar):</p>
 *
 * <pre>{@code
 * const picker = new WidgetPicker({
 *     entries     : pickerEntries,           // already filtered (PINNED removed)
 *     disabledIds : { "MyWidget": true },    // optional — greys those tiles
 *     onPick      : function (entry, params) { ... },
 *     onCancel    : function () { ... }
 * });
 * picker.mountInto(hostEl);
 * }</pre>
 *
 * <p>Entry data comes from {@link WidgetEntriesJson} (a JSON array
 * emitted into the page by {@link WorkspaceShell#widgetEntriesJson()} —
 * one entry per registered {@link WidgetEntry} with display metadata,
 * lifecycle hint, dynamic-import module URL, and Params record schema).
 * The chrome filters PINNED entries out of {@code pickerEntries} (they
 * auto-spawn at boot, never user-addable).</p>
 *
 * <h2>onPick semantics</h2>
 *
 * <ul>
 *   <li>{@code onPick(entry, params)} with a non-null {@code params}:
 *       user confirmed a widget; {@code params} is the value map (empty
 *       object for paramless widgets, or values from the form).</li>
 *   <li>{@code onPick(entry, null)}: user clicked a disabled tile (e.g.
 *       SINGLETON already-open). Caller decides what to do —
 *       typically focus the existing instance.</li>
 *   <li>{@code onCancel()}: user clicked Cancel in the params form.
 *       Picker self-dismisses; caller cleans up its host (typically
 *       removes the empty picker tab).</li>
 * </ul>
 *
 * <h2>Modal? No.</h2>
 *
 * <p>The picker doesn't construct a Modal anymore. Earlier iterations
 * (b.2a / b.2b) had the picker open its own Modal then stage the
 * picked widget in a second Modal that the user dragged to dock; that
 * flow was retired in b.2c for per-pane "+" + inline rendering. The
 * Modal primitive still earns its keep for the drag controller's
 * chip-detach gesture, just not here.</p>
 *
 * <p>See {@code Rfc0025Ext1Doc.md} Mechanism 2 Decisions section for
 * the full design journey (b.2a → b.2b → b.2c → b.2d).</p>
 *
 * @since RFC 0025 Ext1b — Mechanism 2 (Widget Selector / Picker)
 */
public record WidgetPickerModule() implements DomModule<WidgetPickerModule> {

    /** The single export — the {@code WidgetPicker} JS class. */
    public record WidgetPicker() implements Exportable._Constant<WidgetPickerModule> {}

    public static final WidgetPickerModule INSTANCE = new WidgetPickerModule();

    @Override
    public ImportsFor<WidgetPickerModule> imports() {
        return ImportsFor.<WidgetPickerModule>builder()
                .add(new ModuleImports<>(
                        List.of(
                                new WidgetPickerStyles.hwp_grid(),
                                new WidgetPickerStyles.hwp_group_label(),
                                new WidgetPickerStyles.hwp_tile(),
                                new WidgetPickerStyles.hwp_tile_disabled(),
                                new WidgetPickerStyles.hwp_tile_icon(),
                                new WidgetPickerStyles.hwp_tile_label(),
                                new WidgetPickerStyles.hwp_tile_desc(),
                                new WidgetPickerStyles.hwp_form(),
                                new WidgetPickerStyles.hwp_form_row(),
                                new WidgetPickerStyles.hwp_form_label(),
                                new WidgetPickerStyles.hwp_form_input(),
                                new WidgetPickerStyles.hwp_form_actions(),
                                new WidgetPickerStyles.hwp_form_btn(),
                                new WidgetPickerStyles.hwp_form_btn_primary()),
                        WidgetPickerStyles.INSTANCE))
                .build();
    }

    @Override
    public ExportsOf<WidgetPickerModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new WidgetPicker()));
    }
}
